### PR TITLE
Feature: Added hashes tab to properties window

### LINKED ISSUE #3722: Feature: Add hashes tab to the properties window
### Describe the solution
[HashTab](http://implbits.com/products/hashtab/) is a plugin of Windows Explorer.

image

I want the same feature that build in Files.

### Requirements
- Add hashes tab to the properties windows
- Display hashes (starting with CRC32, MD5 and SHA-1) in the list (should follow the design pattern in the details tab)

### Files version

2.3.12.0

### Windows version

10.0.19043.2006

### PR DESCRIPTION
**Pull request by 0x5bfa**
### Description

I'm adding hash page to storage object's properties window.

### Motivation and Context

- Closes #3722
- [x] Add hashes page
  - [x] Calculate hashes and display
  - [x] Add copy button
  - [x] Display ProgressRing when loading
  - [x] Display TextBlock that show 'task was failed'
- [x] Fix NavigationViewStoreStyle

### Validation

- [x] Built and ran the app
- [x] Tested the changes for accessibility

### Screenshots

Loading
image

Left: Managed to load, Right: Failed to load
image

### CODE REVIEW

File: src/Files.App/Strings/en-US/Resources.resw [hunk 1]
```diff
@@ -2607,4 +2607,7 @@
   <data name="MultiSelect" xml:space="preserve">
     <value>Multiselect</value>
   </data>
+  <data name="Hash" xml:space="preserve">
+    <value>Hash</value>
+  </data>
```
Comment (yaira2) [inline src/Files.App/Strings/en-US/Resources.resw:2612]:
```suggestion
  <data name="Hashes" xml:space="preserve">
    <value>Hashes</value>
  </data>
```

File: src/Files.App/Views/Properties/MainPropertiesPage.xaml.cs [hunk 2]
```diff
@@ -185,6 +189,13 @@ private void AddNavigationViewItemsToControl(object item)
 				OutlinePathIcon = (string)Application.Current.Resources["ShieldIconRegular"],
 				FilledPathIcon = (string)Application.Current.Resources["ShieldIconFilled"],
 			};
+			var hashItem = new SquareNavViewItem()
+			{
+				Name = "Hash".GetLocalizedResource(),
+				ItemType = PropertyNavigationViewItemEnums.ItemHash,
+				OutlinePathIcon = (string)Application.Current.Resources["WindowsAppsRegular"],
```
Comment (yaira2) [inline src/Files.App/Views/Properties/MainPropertiesPage.xaml.cs:196]: I'll work on the icons for this

**Comment by onein528** [comment]
*Please don't review until this'd be marked as RFR, cuz there're some ridiculous codes.*

File: src/Files.App/Strings/en-US/Resources.resw [hunk 3]
```diff
@@ -2607,4 +2607,10 @@
   <data name="MultiSelect" xml:space="preserve">
     <value>Multiselect</value>
   </data>
+  <data name="Hashes" xml:space="preserve">
+    <value>Hashes</value>
+  </data>
+  <data name="HashesUnavailable" xml:space="preserve">
+    <value>Hashes are unavailable</value>
```
Comment (yaira2) [inline src/Files.App/Strings/en-US/Resources.resw:2614]: When will this show?

Comment (onein528) [reply src/Files.App/Strings/en-US/Resources.resw:2614]: When file cannot be accessed == When File.Open() throw an exception

Comment (yaira2) [reply src/Files.App/Strings/en-US/Resources.resw:2614]: Shouldn't the tab be hidden?

Comment (onein528) [reply src/Files.App/Strings/en-US/Resources.resw:2614]: That's good idea.

**Comment by yaira2** [comment]
Clicking the hashes page for large files causes the properties window to freeze.

File: src/Files.App/ResourceDictionaries/NavigationViewStoreStyle.xaml [hunk 4]
```diff
@@ -1,12 +1,13 @@
 ﻿<ResourceDictionary
 	xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
 	xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+	xmlns:animatedvisuals="using:Microsoft.UI.Xaml.Controls.AnimatedVisuals"
 	xmlns:primitives="using:Microsoft.UI.Xaml.Controls.Primitives">
```
Comment (ferrariofilippo) [inline src/Files.App/ResourceDictionaries/NavigationViewStoreStyle.xaml:5]: This is not needed

**Review by ferrariofilippo — CHANGES_REQUESTED** [review]
One little thing, otherwise LGTM and works for me

**Comment by onein528** [comment]
I cannot fix an issue where Files will freeze as loading a large file(1 GiB or more).

- CancellationToken doesn't work
- DispatcherQueue doesn't work
- ICommand doesn't work

await is bad? maybe Task.Run()?

**Comment by hishitetsu** [comment]
How about using `HashDataAsync()`? It accepts `CancellationToken`.
https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.sha512.hashdataasync?view=net-7.0

**Comment by mdtauk** [comment]
> I cannot fix an issue where Files will freeze as loading a large file(1 GiB or more).
> 
> * CancellationToken doesn't work
> * DispatcherQueue doesn't work
> * ICommand doesn't work
> 
> await is bad? maybe Task.Run()?

Have you included a progress ring or progress bar as the hashes are being calculated?

**Comment by 0x5bfa** [comment]
@yair100 Tried new one. Can you check also? Works fine for me.

File: src/Files.App/ViewModels/Properties/HashesViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	{
+		public HashesViewModel(ListedItem item)
+		{
+			Item = item;
+			Hashes = new();
+			CanAccessFile = true;
+			CancellationTokenSource = new();
+			LoadAndCalcHashesCommand = new(ExecuteLoadAndCalcHashesCommandAsync);
+
+			//if (LoadFileContent.CanExecute(CancellationTokenSource.Token))
```
Comment (yaira2) [inline src/Files.App/ViewModels/Properties/HashesViewModel.cs:25]: Can this be removed?

File: src/Files.App/ViewModels/Properties/HashesViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+		public CancellationTokenSource CancellationTokenSource { get; set; }
+
+		private bool _isLoading;
+		public bool IsLoading
+		{
+			get => _isLoading;
+			set => SetProperty(ref _isLoading, value);
+		}
+		#endregion
```
Comment (yaira2) [inline src/Files.App/ViewModels/Properties/HashesViewModel.cs:60]:
```suggestion
```

File: src/Files.App/ViewModels/Properties/HashesViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+			Hashes = new();
+			CanAccessFile = true;
+			CancellationTokenSource = new();
+			LoadAndCalcHashesCommand = new(ExecuteLoadAndCalcHashesCommandAsync);
+
+			//if (LoadFileContent.CanExecute(CancellationTokenSource.Token))
+			//	LoadFileContent.Execute(CancellationTokenSource.Token);
+		}
+
+		#region Fields and Properties
```
Comment (yaira2) [inline src/Files.App/ViewModels/Properties/HashesViewModel.cs:29]:
```suggestion
```

File: src/Files.App/Views/Properties/HashesPage.xaml [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+			<TextBlock
+				x:Name="UnavailableTextBlock"
+				Margin="16,56,16,16"
+				x:Load="{x:Bind HashesViewModel.CanAccessFile, Converter={StaticResource BoolNegationConverter}, Mode=OneWay}"
+				Style="{StaticResource BodyTextBlockStyle}"
+				Text="{helpers:ResourceString Name=UnableToCalcHashes}" />
+		</Grid>
+
+		<!--  Loading  -->
+		<StackPanel
```
Comment (yaira2) [inline src/Files.App/Views/Properties/HashesPage.xaml:130]: Can this be moved into the grid above?

File: src/Files.App/Views/Properties/HashesPage.xaml [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+	mc:Ignorable="d">
+
+	<vm:BasePropertiesPage.Resources>
+		<ResourceDictionary>
+			<toolkitconverters:BoolNegationConverter x:Key="BoolNegationConverter" />
+			<toolkitconverters:BoolToVisibilityConverter x:Key="BoolToVisibilityConverter" />
+		</ResourceDictionary>
+	</vm:BasePropertiesPage.Resources>
+
+	<Grid x:Name="RootGrid">
```
Comment (yaira2) [inline src/Files.App/Views/Properties/HashesPage.xaml:22]: I think this grid is extra, the StackPanel can be move to the HashesListGird.

Comment (onein528) [reply src/Files.App/Views/Properties/HashesPage.xaml:22]: HashesListGird has border and background(like a card). Even though I make ListView have border and background, I cannot put UnavailableTextBlock inside ListView. so not possible.

Comment (yaira2) [reply src/Files.App/Views/Properties/HashesPage.xaml:22]: 👍

**Comment by yaira2** [comment]
> Have you included a progress ring or progress bar as the hashes are being calculated?

@mdtauk there is a progress ring 👍

**Review by yair100 — APPROVED** [review]
LGTM

**Pull request merged**